server/src/main/java/io/druid/query/ResultLevelCachingQueryRunner.java

```diff
@@ -0,0 +1,301 @@
+public class ResultLevelCachingQueryRunner<T> implements QueryRunner<T>
+{
+  private static final Logger log = new Logger(ResultLevelCachingQueryRunner.class);
+  private final QueryRunner baseRunner;
+  private ObjectMapper objectMapper;
+  private final Cache cache;
+  private final CacheConfig cacheConfig;
+  private final boolean useResultCache;
+  private final boolean populateResultCache;
+  private Query<T> query;
+  private final CacheStrategy<T, Object, Query<T>> strategy;
+
+
+  public ResultLevelCachingQueryRunner(
+      QueryRunner baseRunner,
+      QueryToolChest queryToolChest,
+      Query<T> query,
+      ObjectMapper objectMapper,
+      Cache cache,
+      CacheConfig cacheConfig
+  )
+  {
+    this.baseRunner = baseRunner;
+    this.objectMapper = objectMapper;
+    this.cache = cache;
+    this.cacheConfig = cacheConfig;
+    this.query = query;
+    this.strategy = queryToolChest.getCacheStrategy(query);
+    this.populateResultCache = ResultLevelCacheUtil.populateResultLevelCacheOnBrokers(query, strategy, cacheConfig);
+    this.useResultCache = ResultLevelCacheUtil.useResultLevelCacheOnBrokers(query, strategy, cacheConfig);
+  }
+
+  @Override
+  public Sequence<T> run(QueryPlus queryPlus, Map responseContext)
+  {
+    if (useResultCache) {
+
+      final String cacheKeyStr = StringUtils.fromUtf8(strategy.computeCacheKey(query));
+      final byte[] cachedResultSet = fetchResultsFromResultLevelCache(cacheKeyStr);
+      String existingResultSetId = extractEtagFromResults(cachedResultSet);
+
+      existingResultSetId = existingResultSetId == null ? "" : existingResultSetId;
+
+      if (query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH) == null) {
+        query = query.withOverriddenContext(
+            ImmutableMap.of(QueryResource.HEADER_IF_NONE_MATCH, existingResultSetId));
+      }
+      Sequence<T> resultFromClient = baseRunner.run(
+          QueryPlus.wrap(query),
+          responseContext
+      );
+      String newResultSetId = (String) responseContext.get(QueryResource.HEADER_ETAG);
+
+      if (newResultSetId != null && newResultSetId.equals(existingResultSetId)) {
+        log.debug("Return cached result set as there is no change in identifiers for query %s ", query.getId());
+        return deserializeResults(cachedResultSet, strategy, existingResultSetId);
+      } else {
+        @Nullable
+        ResultLevelCachePopulator resultLevelCachePopulator = createResultLevelCachePopulator(
+            cacheKeyStr,
+            newResultSetId
+        );
+        if (resultLevelCachePopulator == null) {
+          return resultFromClient;
+        }
+        final Function<T, Object> cacheFn = strategy.prepareForCache(true);
+
+        return Sequences.wrap(Sequences.map(
+            resultFromClient,
+            new Function<T, T>()
+            {
+              @Override
+              public T apply(T input)
+              {
+                if (resultLevelCachePopulator.isShouldPopulate()) {
```

Comment:
we are doing this check for every row in the result set whereas it should only be done once.

Reply:
nmnd, i see its checking whether we've crossed the size limit